File: backend/app/models/mistake.py
import uuid
from datetime import datetime
from sqlalchemy import String, Text, Integer, Boolean, DateTime, ARRAY, Float
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.database import Base


class Mistake(Base):
    __tablename__ = "mistakes"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[str] = mapped_column(String(128), index=True)
    subject: Mapped[str] = mapped_column(String(64))
    question_text: Mapped[str] = mapped_column(Text, default="")
    answer_text: Mapped[str] = mapped_column(Text, default="")
    mistake_reason: Mapped[str] = mapped_column(String(32), default="")
    difficulty: Mapped[int] = mapped_column(Integer, default=3)
    tags: Mapped[list] = mapped_column(ARRAY(String), default=list)
    image_url: Mapped[str] = mapped_column(String(512), default="")
    source: Mapped[str] = mapped_column(String(32), default="")
    correct_count: Mapped[int] = mapped_column(Integer, default=0)
    next_review_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    repetitions: Mapped[int] = mapped_column(Integer, default=0)
    ef: Mapped[float] = mapped_column(Float, default=2.5)
    interval_days: Mapped[int] = mapped_column(Integer, default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.now)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.now, onupdate=datetime.now)

File: frontend/lib/core/models/mistake.dart
class Mistake {
  final String id;
  final String subject;
  final String grade;
  final String questionText;
  final String answerText;
  final String mistakeReason;
  final int difficulty;
  final List<String> tags;
  final String? imageUrl;
  final String source;
  final int correctCount;
  final DateTime? nextReviewAt;
  final DateTime createdAt;

  Mistake({
    required this.id,
    required this.subject,
    this.grade = '',
    this.questionText = '',
    this.answerText = '',
    this.mistakeReason = '',
    this.difficulty = 3,
    this.tags = const [],
    this.imageUrl,
    this.source = '',
    this.correctCount = 0,
    this.nextReviewAt,
    DateTime? createdAt,
  }) : createdAt = createdAt ?? DateTime.now();

  factory Mistake.fromJson(Map<String, dynamic> json) => Mistake(
    id: json['id'] as String,
    subject: json['subject'] as String? ?? '',
    grade: json['grade'] as String? ?? '',
    questionText: json['question_text'] as String? ?? '',
    answerText: json['answer_text'] as String? ?? '',
    mistakeReason: json['mistake_reason'] as String? ?? '',
    difficulty: json['difficulty'] as int? ?? 3,
    tags: List<String>.from(json['tags'] ?? []),
    imageUrl: json['image_url'] as String?,
    source: json['source'] as String? ?? '',
    correctCount: json['correct_count'] as int? ?? 0,
    nextReviewAt: json['next_review_at'] != null ? DateTime.parse(json['next_review_at']) : null,
    createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : DateTime.now(),
  );

  Map<String, dynamic> toJson() => {
    'subject': subject,
    'grade': grade,
    'question_text': questionText,
    'answer_text': answerText,
    'mistake_reason': mistakeReason,
    'difficulty': difficulty,
    'tags': tags,
    'source': source,
  };
}
